Add Eager Options Validation: ValidateOnStart API

maryamariyan opened the pull request:
- Supports Options Validation, getting triggered at startup via a hosted service.
- Allows options validation for named options as well.

**Note:** Microsoft.Extensions.Options would need a new dependency to Microsoft.Extensions.Hosting.Abstractions because we use IHostedService to implement this feature.

Minor TODO:

- [x] Improve unit tests + code coverage
- [x] Get API approved
- [x] Cleanup comments

Fixes https://github.com/dotnet/runtime/issues/36391

maryamariyan committed:
Add Options Validation on Start

davidfowl commented on src/libraries/Microsoft.Extensions.Options/src/OptionsBuilderValidationExtensions.cs:
```diff
@@ -0,0 +1,41 @@
+            _ = optionsBuilder ?? throw new ArgumentNullException(nameof(optionsBuilder));
+
+            // This will only add the hosted service once
+            _ = optionsBuilder.Services.AddHostedService<ValidationHostedService>();
```
You can remove the _ here.

davidfowl commented on src/libraries/Microsoft.Extensions.Options/src/ValidationHostedService.cs:
```diff
@@ -0,0 +1,58 @@
+            if (exceptions.Count == 1)
+            {
+                // Rethrow if it's a single error
+                throw exceptions[0];
```
Should this be using `ExceptionDispatchInfo.Throw(exceptions[0])`?

davidfowl commented on src/libraries/Microsoft.Extensions.Options/src/ValidationHostedService.cs:
```diff
@@ -0,0 +1,58 @@
+        {
+            var exceptions = new List<Exception>();
+
+            foreach (var validate in _validators.Values)
```
Don't use `.Values`, that results in an allocation.

davidfowl commented on src/libraries/Microsoft.Extensions.Options/src/ValidatorOptions.cs:
```diff
@@ -0,0 +1,20 @@
+        // The key maps to the TOptions in IOptions<TOptions> and the value is a method
+        // that accesses the IOptions<TOptions>.Value property in order to force evaluation of
+        // the options type.
+        public ConcurrentDictionary<Type, Action> Validators { get; } = new ConcurrentDictionary<Type, Action>();
```
Why is this concurrent?

maryamariyan commented on src/libraries/Microsoft.Extensions.Options/src/OptionsBuilderValidationExtensions.cs:
```diff
@@ -0,0 +1,41 @@
+        /// <param name="optionsBuilder">The <see cref="OptionsBuilder{TOptions}"/> to configure options instance.</param>
+        /// <returns>The <see cref="OptionsBuilder{TOptions}"/> so that additional calls can be chained.</returns>
+        public static OptionsBuilder<TOptions> ValidateOnStart<TOptions>(this OptionsBuilder<TOptions> optionsBuilder)
+            where TOptions : class, new()
```
@davidfowl do we want this new() constraint?

davidfowl replied:
Nope

maryamariyan committed:
Apply PR feedback:

- Drop new() constraint
- Remove _ in most cases
- Remove the ConcurrentDictionary
- Slight comment cleanup
- Skip tests on mono: due to PNSE on host.StartAsync()

maryamariyan committed:
Cleanup test class

maryamariyan commented on src/libraries/Microsoft.Extensions.Options/tests/Microsoft.Extensions.Options.Tests/OptionsBuilderValidationTests.cs:
```diff
@@ -0,0 +1,453 @@
+                    })
+                    .Validate(o => o.Boolean, "Boolean")
+                    .Validate(o => o.Integer > 12, "Integer")
+                    .ValidateOnStart();
```
Because of

https://github.com/dotnet/runtime/blob/68e1b66345d8e4d8174f61c892201f76a18331b9/src/libraries/Microsoft.Extensions.Options/src/OptionsBuilderValidationExtensions.cs#L34

If we modified `CreateHostBuilder` in this test to have a correct configuration as its last ValidateOnStart options setup, then we get no failures on startup at all. (check below) 

```c#


            var hostBuilder = CreateHostBuilder(services =>
            {
                services.AddOptions<ComplexOptions>("bad_configuration1")
                    .Configure(o => o.Boolean = false)
                    .Validate(o => o.Boolean, "bad_configuration1")
                    .ValidateOnStart();

                services.AddOptions<ComplexOptions>("bad_configuration2")
                    .Configure(o =>
                    {
                        o.Boolean = false;
                        o.Integer = 11;
                    })
                    .Validate(o => o.Boolean, "Boolean")
                    .Validate(o => o.Integer > 12, "Integer")
                    .ValidateOnStart();

                services.AddOptions<ComplexOptions>("correct_configuration")
                    .Configure(o => o.Boolean = true)
                    .Validate(o => o.Boolean, "Boolean must be true.")
                    .ValidateOnStart();
            });
```

davidfowl replied:
That's by design. Validation runs after constructing the final object by running all of the delegates. Validation is attached to the TOptions type not the specific call to AddOptions

maryamariyan commented on src/libraries/Microsoft.Extensions.Options/src/OptionsBuilderValidationExtensions.cs:
```diff
@@ -0,0 +1,40 @@
+                {
+                    // This adds an action that resolves the options value to force evaluation
+                    // We don't care about the result as duplicates are not important
+                    vo.Validators[typeof(TOptions)] = () => options.Get(optionsBuilder.Name);
```
check comment: https://github.com/dotnet/runtime/pull/47821/files#r570699559

maryamariyan commented on src/libraries/Microsoft.Extensions.Options/src/OptionsBuilderValidationExtensions.cs:
```diff
@@ -0,0 +1,40 @@
+    /// <summary>
+    /// Extension methods for adding configuration related options services to the DI container via <see cref="OptionsBuilder{TOptions}"/>.
+    /// </summary>
+    public static class OptionsBuilderValidationExtensions
```
The word `Validation` can be dropped for this class, (suggesting rename to `OptionsBuilderExtensions`).
This allows any future extensions methods (non-validation specific) to also make sense to be added here.

Just awaiting approval from .NET API reviewers before changing this on the PR.

maryamariyan committed:
Rename

eerhardt commented on src/libraries/Microsoft.Extensions.Options/src/Microsoft.Extensions.Options.csproj:
```diff
@@ -14,6 +14,7 @@
 
   <ItemGroup>
     <ProjectReference Include="$(LibrariesProjectRoot)Microsoft.Extensions.DependencyInjection.Abstractions\src\Microsoft.Extensions.DependencyInjection.Abstractions.csproj" />
+    <ProjectReference Include="$(LibrariesProjectRoot)Microsoft.Extensions.Hosting.Abstractions\src\Microsoft.Extensions.Hosting.Abstractions.csproj" />
```
Do we really want this layering? Can't I use Options without Hosting?

davidfowl replied:
Should it be in the hosting assembly?

eerhardt replied:
Why do we need a hosted service at all? That’s the part I’m not understanding

davidfowl replied:
Yes, the goal of this feature is to run validation code on start. That can only be done if something resolved the options value. In that case, that thing doing the resolution is the hosted service.

eerhardt replied:
But only on "start" of a Host? Is there a way to get code executed right after the DI container is built?

I'm concerned with putting this feature into the "Options" assembly, but it doesn't work if I just use "Options" and I don't use Hosting.

davidfowl replied:
This only works for the hosting model, there's no other way to "run code at startup". Maybe this should be in hosting.

eerhardt replied:
DI doesn't have a hook that I can run some code after the container is built?

davidfowl replied:
No, it doesn't have that. But even if it did, we can't have options tied to the specific our container implementation. The other feature request we have gotten though is to allow pre-warming singletons. That's something that could be in the container. But it wouldn't help here, you'd need to add another singleton that resolved the options type in the constructor (which I guess could work?)

maryamariyan replied:
#### Regarding assembly:
Based on the above discussion it makes sense to move the diff onto the hosting assembly. 

#### Regarding namespace:
The namespace for ValidateOnStart was approved under Microsoft.Extensions.DependencyInjection in the .NET API review due to reasons:

1. because it is used along with configuring the IServiceCollection. In the ASP.NET Core templates' Startup.cs classes we have a using for Microsoft.Extensions.DependencyInjection but not Microsoft.Extensions.Options
2. keeping consistent with the most common APIs related to OptionsBuilder:

https://github.com/dotnet/runtime/blob/1d9e50cb4735df46d3de0cee5791e97295eaf588/src/libraries/Microsoft.Extensions.Options.DataAnnotations/ref/Microsoft.Extensions.Options.DataAnnotations.cs#L11

https://github.com/dotnet/runtime/blob/1d9e50cb4735df46d3de0cee5791e97295eaf588/src/libraries/Microsoft.Extensions.Options/ref/Microsoft.Extensions.Options.cs#L12-L13

https://github.com/dotnet/runtime/blob/1d9e50cb4735df46d3de0cee5791e97295eaf588/src/libraries/Microsoft.Extensions.Options.ConfigurationExtensions/ref/Microsoft.Extensions.Options.ConfigurationExtensions.cs#L11-L13

cc: @halter73

maryamariyan committed:
Move to hosting assembly

maryamariyan committed:
- Locate tests properly:
Keeps DataAnnotations tests in Options and rest in Hosting

maryamariyan committed:
fix compile

maryamariyan commented:
The CI failure is unrelated.

@davidfowl @eerhardt The PR should be ready for another review.

eerhardt commented on src/libraries/Microsoft.Extensions.Options/tests/Microsoft.Extensions.Options.Tests/Microsoft.Extensions.Options.Tests.csproj:
```diff
@@ -8,6 +8,7 @@
   <ItemGroup>
     <ProjectReference Include="$(LibrariesProjectRoot)Microsoft.Extensions.Configuration\src\Microsoft.Extensions.Configuration.csproj"/>
     <ProjectReference Include="$(LibrariesProjectRoot)Microsoft.Extensions.DependencyInjection\src\Microsoft.Extensions.DependencyInjection.csproj" />
+    <ProjectReference Include="$(LibrariesProjectRoot)Microsoft.Extensions.Hosting\src\Microsoft.Extensions.Hosting.csproj"/>
```
Since the code the new tests are testing is in Hosting, would it make sense to move these tests to the Hosting tests?

maryamariyan replied:
I am only keeping DataAnnotation tests in Options. The rest are in hosting.

eerhardt commented on src/libraries/Microsoft.Extensions.Options/tests/Microsoft.Extensions.Options.Tests/OptionsBuilderTest.cs:
```diff
@@ -665,5 +665,125 @@ public void CanValidateMixDataAnnotations()
         }
+
+        [Fact]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/34580", TestPlatforms.Windows, TargetFrameworkMonikers.Netcoreapp, TestRuntimes.Mono)]
```
This issue is closed as a dupe of #34582. Can we update the link to point to the active issue?

maryamariyan replied:
I'll address this in another PR there are plenty of tests currently linked with the dupe.

eerhardt commented on src/libraries/Microsoft.Extensions.Options/tests/Microsoft.Extensions.Options.Tests/OptionsBuilderTest.cs:
```diff
@@ -665,5 +665,125 @@ public void CanValidateMixDataAnnotations()
+
+            try
+            {
+                var value = sp.GetRequiredService<IOptions<AnnotatedOptions>>().Value;
```
Is there something about `value` that we can assert? At least that it is non-null? Or anything else?

eerhardt commented on src/libraries/Microsoft.Extensions.Options/tests/Microsoft.Extensions.Options.Tests/OptionsBuilderTest.cs:
```diff
@@ -665,5 +665,125 @@ public void CanValidateMixDataAnnotations()
+                var value = sp.GetRequiredService<IOptions<AnnotatedOptions>>().Value;
+            }
+#pragma warning disable CA1031 // Do not catch general exception types
+            catch (Exception ex)
```
Why are we catching and asserting here? Why not just take the `try` out and let the test throw if it does?

eerhardt commented on src/libraries/Microsoft.Extensions.Hosting/src/OptionsBuilderExtensions.cs:
```diff
@@ -0,0 +1,40 @@
+    public static class OptionsBuilderExtensions
+    {
+        /// <summary>
+        /// Enforces options validation check on start rather then in runtime.
```
`rather then` => `rather than`

eerhardt commented on src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/OptionsBuilderExtensionsTests.cs:
```diff
@@ -0,0 +1,312 @@
+        [Fact]
+        public void ValidateOnStart_NullOptionsBuilder_Throws()
+        {
+            var hostBuilder = CreateHostBuilder(services =>
```
Is this needed? Why not just call

`OptionsBuilderExtensions.ValidateOnStart(null)`?

eerhardt commented on src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/OptionsBuilderExtensionsTests.cs:
```diff
@@ -0,0 +1,312 @@
+                services.AddOptions<ComplexOptions>()
+                    .Configure(o => o.Boolean = false)
+                    .Validate(o => o.Boolean, "first Boolean must be true.")
+                    .ValidateOnStart();
```
Do we have any tests that call `.ValidateOnStart()` with valid options and make sure no errors are raised?

maryamariyan replied:
Since the tests are now sectioned into two different places, I'm gonna look back at the `..Hosting/tests/UnitTests/OptionsBuilderExtensionsTests.cs` test file and make sure it contains all the unit tests we want and leave the other one in `..Options.Tests/OptionsBuilderTest.cs` to focus solely on DataAnnotations.

maryamariyan committed:
Apply PR feedback. Improve unit tests

eerhardt commented on src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/OptionsBuilderExtensionsTests.cs:
```diff
@@ -0,0 +1,359 @@
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/34580", TestPlatforms.Windows, TargetFrameworkMonikers.Netcoreapp, TestRuntimes.Mono)]
+        private async Task ValidateOnStart_AddOptionsMultipleTimesForSameType_LastOneGetsTriggered()
+        {
+            bool firstOptionsBuilderIgnored = true;
```
(nit) I would make both of these booleans use the same pattern. To me it is more understandable to have:

```C#
bool firstOptionsBuilderTriggered = false;
bool secondOptionsBuilderTriggered = false;


Assert.False(firstOptionsBuilderTriggered);
Assert.True(secondOptionsBuilderTriggered);
```

eerhardt reviewed (approved):
Looks good. Thanks for driving this through, @maryamariyan.

davidfowl reviewed (approved):
Still want to discuss the home

maryamariyan commented:
> Still want to discuss the home

OK, marking unresolved the conversation thread above: https://github.com/dotnet/runtime/pull/47821#discussion_r573383024

Here's a summary of the two approaches tried:

- [ ] `Microsoft.Extensions.Options` (would need a reference to `Microsoft.Extensions.Hosting.Abstractions` to access IHostedService for implementing `ValidateOnStart`
- [ ] `Microsoft.Extensions.Hosting` (in PR as-is, it already references `M.E.Options` and `M.E.Hosting.Abstractions`)

maryamariyan committed:
Apply nit feedback

halter73 commented on src/libraries/Microsoft.Extensions.Hosting/src/ValidatorOptions.cs:
```diff
@@ -0,0 +1,14 @@
+    internal class ValidatorOptions
+    {
+        // Maps each options type to a method that forces its evaluation, e.g. IOptionsMonitor<TOptions>.Get(name)
+        public IDictionary<Type, Action> Validators { get; } = new Dictionary<Type, Action>();
```
I know this is an internal implementation detail, but is there any particular reason to use a dictionary over a list of actions for this?

davidfowl replied:
True

eerhardt replied:
My understanding is that for a `TOptions`, we only need to run the `Action` once. If `ValidateOnStart` is called multiple times for the same `TOptions`, we only want it to be "validated" once.

davidfowl replied:
Ah yea thats why I did it 😄

eerhardt replied:
@maryamariyan - it probably isn't worth resetting CI ... but since I had the same question initially, maybe add a comment here explaining why a Dictionary is used next time you are touching this code.
